app/main_dev/launch.js

```diff
@@ -335,6 +371,125 @@ export const launchDCRD = (reactIPC, testnet, appdata) =>
+// not defined, we read dcrd.conf file from our app directory. If it does not
+// exist, a new conf file is created with random rpc_user and rpc_pass.
+function readDcrdConfig(testnet, appdata) {
+  // createTempDcrdConf creates a temp dcrd conf file and writes it
+  // to decreditons config directory: getAppDataDirectory()
+  const createTempDcrdConf = (testnet) => {
+    let dcrdConf = {};
+    if (!fs.existsSync(dcrdCfg(getAppDataDirectory()))) {
+      const port = testnet ? "19109" : "9109";
+
+      dcrdConf = {
+        "Application Options": {
+          rpcuser: makeRandomString(10),
+          rpcpass: makeRandomString(10),
+          rpclisten: `127.0.0.1:${port}`
+        }
+      };
+      fs.writeFileSync(dcrdCfg(getAppDataDirectory()), ini.stringify(dcrdConf));
+    }
+    return getAppDataDirectory();
+  };
+
+  try {
+    let rpc_host = "127.0.0.1";
+    let rpc_port = testnet ? "19109" : "9109";
+    let dcrdAppdata = appdata;
+    let rpc_cert, rpc_key, configFile, rpc_user, rpc_pass;
+
+    // if appdata is set, it means it is an advanced start, so we need to
+    // warn about the need to update the rpc.cert and rpc.key files. Else
+    // we use decrediton's own rpc.cert and rpc.key.
+    if (appdata) {
+      dcrdAppdata = appdata;
+      rpc_cert = `${appdata}/rpc.cert`;
+      rpc_key = `${appdata}/rpc.key`;
+      // if conf file of the appdata informed exists, we use it, otherwise
+      // we use decrediton's own dcrd.conf file.
+      if (fs.existsSync(dcrdCfg(appdata))) {
+        configFile = appdata;
+      } else {
+        configFile = getAppDataDirectory();
+      }
+    } else {
+      rpc_cert = `${getAppDataDirectory()}/rpc.cert`;
+      rpc_key = `${getAppDataDirectory()}/rpc.key`;
+      dcrdAppdata = getDcrdPath();
+      configFile = getAppDataDirectory();
+    }
+
+    // if dcrd.conf file is from decrediton dir and does not exist we create a
+    // new one.
+    if (
+      configFile === getAppDataDirectory() &&
+      !fs.existsSync(dcrdCfg(configFile))
+    ) {
+      createTempDcrdConf(testnet);
+    }
+    const readCfg = ini.parse(
+      Buffer.from(fs.readFileSync(dcrdCfg(configFile))).toString()
+    );
+
+    let userFound,
+      passFound = false;
+    // Look through all top level config entries
+    for (const [key, value] of Object.entries(readCfg)) {
+      if (key === "rpcuser") {
+        rpc_user = value;
+        userFound = true;
+      }
+      if (key === "rpcpass") {
+        rpc_pass = value;
+        passFound = true;
+      }
+      if (key === "rpclisten") {
+        const splitListen = value.split(":");
+        if (splitListen.length >= 2) {
+          rpc_host = splitListen[0];
+          rpc_port = splitListen[1];
+        }
+      }
+      if (!userFound && !passFound) {
+        // If user and pass aren't found on the top level, look through all
+        // next level config entries
+        for (const [key2, value2] of Object.entries(value)) {
```

Comment:
I'm confused: what sort of ini file requires parsing like this?

Reply:
That's a good question, tbh I only copied the code from our config.js to here. 

This was implement by @alexlyp  on https://github.com/decred/decrediton/pull/1111 maybe he recalls?

Reply:
Don't remember the specifics, but I believe there were some corner cases that weren't allowing for the user/pass to be plucked from the config files properly.  So if they weren't found on the top level, then they would be looked through all the key/value config pairs for any hits.